File: problems/python/src/util/dicts.py
from bisect import bisect_right
from typing import TypeVar, Any

T = TypeVar('T')


def floor_key(dictionary: dict[T, Any], target: T) -> T | None:
    keys = sorted(dictionary.keys())
    index = bisect_right(keys, target)
    if index > 0:
        return keys[index - 1]
    return None

File: javascript/src/util/dicts.js
export function floorKey(dictionary, target) {
  const keys =
      Object
          .keys(dictionary)
          .map(key => +key)
          .sort((a, b) => a < b ? -1 : a > b ? 1 : 0);
  const index = _bisectRight(keys, target);
  return index > 0
      ? keys[index - 1]
      : undefined;
}

function _bisectRight(arr, target) {
  let lo = 0;
  let hi = arr.length;
  while (lo < hi) {
    const mid = Math.floor((lo + hi) / 2);
    if (arr[mid] <= target) {
      lo = mid + 1;
    } else {
      hi = mid;
    }
  }
  return lo;
}
